apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/checkNoCapitalizedFields.kt

```diff
@@ -0,0 +1,85 @@
+package com.apollographql.apollo3.compiler
+
+import com.apollographql.apollo3.annotations.ApolloInternal
+import com.apollographql.apollo3.ast.GQLDefinition
+import com.apollographql.apollo3.ast.GQLField
+import com.apollographql.apollo3.ast.GQLFragmentDefinition
+import com.apollographql.apollo3.ast.GQLFragmentSpread
+import com.apollographql.apollo3.ast.GQLInlineFragment
+import com.apollographql.apollo3.ast.GQLOperationDefinition
+import com.apollographql.apollo3.ast.GQLSelection
+import com.apollographql.apollo3.ast.Issue
+import com.apollographql.apollo3.ast.internal.IssuesScope
+
+@ApolloInternal
+fun findCapitalizedFields(definitions: List<GQLDefinition>): List<Issue> {
+  val scope = object : ValidationScope {
+    override val issues = mutableListOf<Issue>()
+    override val fragmentsByName = definitions.filterIsInstance<GQLFragmentDefinition>().associateBy { it.name }
+  }
+
+  definitions.forEach {
+    when (it) {
+      is GQLOperationDefinition -> scope.findCapitalizedFields(it.selectionSet.selections)
+      is GQLFragmentDefinition -> scope.findCapitalizedFields(it.selectionSet.selections)
+    }
+  }
+
+  return scope.issues
+}
+
+/**
+ * Fields named with a capital first letter clash with the corresponding model name, unless flatten.
+ */
+private fun ValidationScope.findCapitalizedFields(selections: List<GQLSelection>) {
+  selections.forEach {
+    when (it) {
+      is GQLField -> {
+        val alias = it.alias
+        if (alias != null) {
+          if (isFirstLetterUpperCase(alias)) {
+            issues.add(Issue.UpperCaseField(message = """
+                      Capitalized alias '$alias' is not supported as it causes name clashes with the generated models. Use '${decapitalizeFirstLetter(alias)}' instead.
+                    """.trimIndent(),
+                sourceLocation = it.sourceLocation)
+            )
+          }
+        } else if (isFirstLetterUpperCase(it.name)) {
+          issues.add(Issue.UpperCaseField(message = """
+                      Capitalized field '${it.name}' is not supported as it causes name clashes with the generated models. Use an alias instead or the 'flattenModels' compiler option.
+                    """.trimIndent(),
+              sourceLocation = it.sourceLocation)
+          )
+        }
+        it.selectionSet?.let { selectionSet ->
+          findCapitalizedFields(selectionSet.selections)
+        }
+      }
+      is GQLInlineFragment -> findCapitalizedFields(it.selectionSet.selections)
+      is GQLFragmentSpread -> findCapitalizedFields(fragmentsByName[it.name]!!.selectionSet.selections)
+    }
+  }
+}
+
+private interface ValidationScope : IssuesScope {
+  override val issues: MutableList<Issue>
+  val fragmentsByName: Map<String, GQLFragmentDefinition>
```

Comment:
Random thoughts: It'd be interesting to see if the Kotlin 1.7 context receivers can make the whole "scopes" thing a bit easier to diggest